### ISSUE TITLE
Enhanced Open Source project infrastructure.

### ISSUE THREAD

**Issue by a11n** [taskadapter/redmine-java-api#176]
Beyond doubt, this is a cool project! :-) But what's missing compared to other great open source projects is a more advanced/automated project infrastructure. The following would be nice to have:

- [GitHub Project Page](https://pages.github.com/#project-site)
- [Automatic builds & tests](http://docs.travis-ci.com) (with a nice "build:passed" badge ;))
- Auto-deployment of releases
- Javadoc (auto-generated and hosted on project page)

I am willing to contribute one ore more of aforementioned infrastructure enhancements. Since some of them require kind of "administrative privileges" (e.g. activating this repo on Travis CI) I want to check first if you also see the need for those enhancements :-) If so, I will make according pull requests.

Examples:
In my fork I already integrated automatic Javadoc generation and deployment:
https://github.com/a11n/redmine-java-api
https://a11n.github.io/redmine-java-api/javadoc/

What do you think?

**Comment by alexeyOnGitHub** [comment]
we do have a Jenkins server with automated builds, of course. I don't know if using Travis CI has any advantages. it is nice to use some "cloud" infrastructure, but we still need to host test Redmine, so not much value in using their Jenkins or whatever they have.
failed/passed tests mark - may be a pretty feature, but I don't know how useful that is. tests for code in "master" should always pass. dev team gets mail notification from Jenkins when they fail, so...

auto-deployment of releases - do you mean to Maven Central? there is a build task in Jenkins for that as well, although it is intentionally made manual.

publishing javadoc - that would be useful I think.

**Comment by a11n** [comment]
Hi Alexey,

And first of all many thanks for your reply :-)

It’s good to know, that you are automating a lot of stuff (in the background) already. I was not aware of this. Is it documented somewhere? Maybe I’ve overseen it. Is your Jenkins public?

In regards to „open development“ this cloud CIs do a great job, they integrate well with GitHub and their build logs are public for everybody (see for instance [1]), which is a useful development artifact IMHO. It’s really worth to have a look, for example at these popular Java libs (notice the .travis.yml) [2][3][4].

If you prefer manual deployment to a central repository this is fine by me :-)
Just wanted to suggest that there is also a way to automate this (and having everything necessary for it in this GH repo, so no side infrastructure needed and everything public to the open source community).

Last but not least, Javadocs. This was the original reason I came up with my suggestions. I found your Redmine API via Google and liked it. Although there is a pretty nice example in the README, it would have helped me (I assume others as well) to have and additional documentation. I see, in this regards you agree with me :-)

I could contribute a shell script for the generation and deployment (to GH pages) via pull request if you like. This will most probably work with Jenkins as well. 

Just that you get me right, I came across your nice project, saw something I would like to have, so I offer to contribute it :-) That’s just my interpretation of open source.

Cheers,
André

[1] https://travis-ci.org/square/javapoet/jobs/52286387
[2] https://github.com/mockito/mockito
[3] https://github.com/jayway/rest-assured
[4] https://github.com/square/okhttp

**Comment by alexeyOnGitHub** [comment]
hi! here is the public Jenkins for the library: https://76.126.10.142:8093 (you can see logs, changes, etc). I do not think this URL is currently documented anywhere. Good point, we can fix this.

About using public infrastructure - it would be nice to unload all the stuff I have here to some public server(s), but... TravisCI is useful for build jobs, but it would not help with hosting several test Redmine instances.

The release deployment is made manual (which means manually clicking a button in one of the Jenkins jobs really) to prevent accidental version change and deployment before the code is ready. we release new versions only once in a few months, so it is not important for this particular project. 

Publishing Javadoc would be useful. Also, I see the documentation we have now is pretty limited (as always...). I use this library in a commercial project and I know it inside out, so I never needed detailed javadocs.. but of course we should go over the code and make sure it is easy to use even for people who see it for the first time.
One thing I would note about javadoc - I don't think we should publish ".internal.*: packages. There is little value in that and it would just clutter the documentation.

Thank you for your help.

**Comment by a11n** [comment]
>I use this library in a commercial project and I know it inside out, so I never needed detailed javadocs..

I know exactly what you are talking about :-)

>but of course we should go over the code and make sure it is easy to use even for people who see it for the first time.

Let me know if you need help. Just choose some "straightforward" classes/methods and I could add the documentation :-)

>One thing I would note about javadoc - I don't think we should publish ".internal.*: packages. There is little value in that and it would just clutter the documentation.

Good hint. I excluded the internal package from the doc generation now. Furthermore I found a solution to integrate everything in the build script, so know `gradle generateJavadocAndPushToGhPages` does all the work.

In regards to an "enhanced open source project infrastructure" I also moved the contribution section from the README file to a dedicated CONTRIBUTION file, which is also good practice on GitHub.

Another thing which I noticed are the release notes. Having them in the Wiki is good, but using a CHANGELOG which is bundled with the repo is also common practice and easier to maintain IMHO. (An it could be auto-generated in the future as well :-))

**Comment by alexeyOnGitHub** [comment]
I think documenting RedmineManagerFactory and RedmineManager would be useful.

I am not a fan of "combined" commands, in my experience splitting them to separate "generate javadoc" and "upload generated javadoc to GH pages" would be easier to (re-)use.

good point about CONTRIBUTING file. whether to have that in a separate wiki page or a file in the repo - I don't see a lot of difference. we can go either way.

**Comment by a11n** [comment]
>I am not a fan of "combined" commands, in my experience splitting them to separate "generate javadoc" and "upload generated javadoc to GH pages" would be easier to (re-)use.
Ok, I'll let those two tasks isolated. :-)

Just had a quick look at RedmineManagerFactory and RedmineManager. The factory class seemed quite good documented to me. The manager appears also sufficient documented to me.

>I think documenting RedmineManagerFactory and RedmineManager would be useful.

I am not sure, if **every** public method has to be documented as long as it is self-describing, like "getIssueManager". Adding documentation "Return the issue manager" won't add any benefit IMHO. But this is a discussion in Java development in general, some prefer full documentation, other prefer a cleaner code. => So, I am not so deep into your code to add any beneficial documentation here :-)

>whether to have that in a separate wiki page or a file in the repo - I don't see a lot of difference. we can go either way.

I think it's a matter of taste :-) I like to have one "single source of truth". Maintaining a CHANGELOG file bundles everything within the repo. It can be updated and pushed along with the changes. There is also a chance to automate the generation of the log. A wiki, on the other hand, is a secondary artifact, which is somehow isolated from the codebase. But, as already mentioned, just personal preference.

**Comment by alexeyOnGitHub** [comment]
right, no need for getters javadoc.

**Comment by a11n** [comment]
Just to keep you informed, I had a deeper look on the plugin I utilized for the GH pages upload and it seems that it only supports user/password, no API key.

So, I suggest to keep things simple at the moment and to generate and push the Javadoc manually at first.

I will set up a PR which configures the Javadoc generation to omit the internal package. I will also document how to generate and push the Javadoc manually. This is of course not the automated approach I had in mind in the first place, but there will be an online Javadoc for this cool API.

**Comment by a11n** [comment]
Ok, I finally got it to work :) (had to struggle with a SHELL scripting for the most of the time...)

I will set up an according PR with documentation.
